docs/design.md

```diff
@@ -0,0 +1,279 @@
+
+## Interfaces
+### Data Provider
+Interface implemented by data providers for the interaction with indexer nodes.
+
+```go
+type IngestionServer interface {
+        // Broadcast advertisement to the network
+        func Broadcast (ctx context.Context, ad Advertisement) error
+        
+        // Push an update for a single entry.
+        // This can be used to perform updates for a small number of CIDs
+        // When a full advertisement is not worth it (web3.storage case).
+        // Indexer may only accept pushes from authenticated providers.
+        func Push (ctx context.Context, indexer peer.ID, entry Entry) 
+}
+```
+
+Data providers need to include protocol handlers able to process and serve indexer node 
+ingestion requests (see [protocol](#protocol)).
+
+### Indexer Node
+Interface implemented by indexer nodes for the interaction with data providers.
+
+```go
+type IngestionClient interface{
+        // Ingest requests the ingestion of a specific range of updates
+        func Ingest (ctx context.Context, p peer.ID, firstID AdvertisementID, lastID AdvertisementID) error
+
+        // Sync with a data provider up to latest ID
+        func Sync (ctx context.Context, p peer.ID, latest AdvertisementID) error
+}
+
+```
+
+Data providers need to include protocol handlers able to process and serve indexer node 
+push Advertisements and immediate ingestions (see [protocol](#protocol)).
+
+## Protocol
+The interaction protocol between data providers and indexer nodes can be divided in two subprotocols: the advertisement protocol, and the ingestion protocol.
+
+## Advertisement Protocol
+It orchestrates the announcement of udpates from data providers to indexer nodes. Every time a data provider receives
+an update for the data they provide (e.g. new deal in a Filecoin miner), they may choose to broadcast an `Advertisement`
+to the network so indexer nodes can start indexing this data. Three kinds of updates will initially be supported:
+- `Pull` Advertisements: Data providers broadcast the `Advertisement` for the new data using a shared
+pub-sub channel. When indexer nodes see these updates, they trigger an ingestion session to sync with the
+data provider.
+- `Push` Advertisements: Some indexer nodes may support advertisement pushes by data providers. Indexers 
+will trigger a new ingestion session to gather the data for that update. 
+- `Immediate` Ingestion: Some data providers may want to share updates for a small number of CIDs which doesn't
+require the generation of a dedicated `Advertisement` and triggering an ingestion session. For this case, indexer
+nodes will provide an endpoint for providers to push small updates directly to them. Indexer nodes may
+restrict immediate ingestion updates to authenticated data providers __(we3.storage case)__.
+
+## Ingestion Protocol
+When indexer nodes realize that they are not in sync with a data provider they are indexing data for, they trigger
```

Comment:
Minor suggestion:  Be specific about how an indexer knows it is not in sync.

"When an indexer sees that a provider has advertised a later index than the indexer currently has..."